### PR TITLE
Improve error reporting for corrupt messages

### PR DESCRIPTION
**Pull request by complexspaces**
Re-introduction and rebasing of https://github.com/rustls/rustls/pull/1016 now that the API is open for breaking changes again.

Closes https://github.com/rustls/rustls/pull/1043

### CODE REVIEW

**Comment by ctz** [comment]
Given there are ten different call sites for `corrupt_message`, maybe we could just describe them in a new `non_exhaustive` enum and put that in `Error::CorruptMessagePayload`?

I haven't seen people putting `core::panic::Location` in error variants in other crates, and I'm not sure what it means in terms of structure and program size?

**Comment by djc** [comment]
Yeah, I think I'd prefer that approach as well, should match well with the Codec yielding Result.

**Comment by complexspaces** [comment]
> maybe we could just describe them in a new non_exhaustive enum and put that in Error::CorruptMessagePayload

Just to double check, are either of you opposed to having ~10 different enum variants to start? One of the goals of this PR was to know exactly which failure branch was reached in the state machine and roughly that many (AFAICT) would be required.

**Comment by djc** [comment]
No, I think Joe was describing exactly that.

**Comment by complexspaces** [comment]
Quick status update: I am still working on this, but I've been short on time. Sorry for the delay.

**Comment by complexspaces** [comment]
I've rebased these changes onto `main` and made the requested changes. As the end result, the existing corrupted message handling also got merged into this "new way" so that there was only one `Error` variant related to corrupt/invalid messages.

Please have a look and let me know if anything else needs fixed or cleaned up.

**Review by djc — COMMENTED** [review]
This is looking great! Here are some further suggestions for improvements.

If you won't have time for this, I'd be happy to contribute further work along these lines.

File: rustls/src/msgs/deframer.rs [hunk 1]
```diff
@@ -75,7 +77,9 @@ impl MessageDeframer {
                 }
                 Err(_) => {
                     self.desynced = true;
-                    return Err(Error::CorruptMessage);
+                    return Err(Error::CorruptMessagePayload(
```
[the remark this answers is not in the document]
Comment (complexspaces) [reply rustls/src/msgs/deframer.rs:80]: Hm, I am not sure I follow yet, sorry. Do you mean that we'd add a new variant to `CorruptMessage` that contains a `MessageError` for the deframer to use and fill with any errors that come out of `OpaqueMessage::read()`? 

> We could potentially rename...

Following, we'd then just rename the variants on `MessageError`?

[the remark this answers is not in the document]
Comment (djc) [reply rustls/src/msgs/deframer.rs:80]: Sorry, I'm proposing that we add the following variants to `CorruptMessage`:

* `HandshakePayloadTooLarge`
* `EmptyApplicationData`
* `MessageTooLarge`
* `InvalidContentType`
* `UnknownProtocolVersion`

And that we modify `MessageError` to have these same variants (from today's `IncorrectSize`, `IllegalLength`, `IllegalContentType` and `IllegalProtocolVersion`), such that we propagate these variants from `MessageError` to the application while keeping `TooShortForHeader` and `TooShortForLength` as is (that is, handled by the deframer).

Does that make sense?

[the remark this answers is not in the document]
Comment (complexspaces) [reply rustls/src/msgs/deframer.rs:80]: I think so, yeah. I take it then that `MessageError` is just another case of [an enum that shouldn't really be exposed from the internals module](https://docs.rs/rustls/latest/rustls/internal/msgs/message/enum.MessageError.html)?

[the remark this answers is not in the document]
Comment (djc) [reply rustls/src/msgs/deframer.rs:80]: Yes!

File: rustls/src/error.rs [hunk 2]
```diff
@@ -1,11 +1,32 @@
...
 use std::error::Error as StdError;
 use std::fmt;
 use std::sync::Arc;
 use std::time::SystemTimeError;
 
+/// A corrupt TLS message payload that resulted in an error.
+#[non_exhaustive]
+#[derive(Debug, Clone)]
+pub enum CorruptMessageError {
+    /// The peer sent us a syntatically incorrect TLS message.
```
[the remark this answers is not in the document]
Comment (complexspaces) [reply rustls/src/error.rs]: Oops, thank you for catching that.

File: rustls/src/error.rs [hunk 3]
```diff
@@ -31,11 +52,8 @@ pub enum Error {
         got_type: HandshakeType,
     },
 
-    /// The peer sent us a syntactically incorrect TLS message.
-    CorruptMessage,
-
     /// The peer sent us a TLS message with invalid contents.
-    CorruptMessagePayload(ContentType),
+    CorruptMessagePayload(CorruptMessageError),
```
[the remark this answers is not in the document]
Comment (complexspaces) [reply rustls/src/error.rs]: I left this one as `CorruptMessage`, even after renaming the inner enum. IMO its really weird to have a variant name and inner type named the exact same thing.

File: rustls/src/msgs/deframer.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (complexspaces) [reply rustls/src/msgs/deframer.rs]: I believe that I've implemented this like discussed, please have a look when you can.

File: rustls/src/msgs/deframer.rs [hunk 4]
```diff
@@ -41,7 +41,9 @@ impl MessageDeframer {
     /// `Ok(Some(_))` if a valid message was found and decrypted successfully.
     pub fn pop(&mut self, record_layer: &mut RecordLayer) -> Result<Option<Deframed>, Error> {
         if self.desynced {
-            return Err(Error::CorruptMessage);
+            return Err(Error::CorruptMessagePayload(
+                CorruptMessageError::IncorrectFrame,
+            ));
```
[the remark this answers is not in the document]
Comment (complexspaces) [reply rustls/src/msgs/deframer.rs]: I took a stab at implementing this in a second commit since it looked fairly simple. Please have a look.

File: rustls/src/error.rs [hunk 5]
```diff
@@ -1,11 +1,32 @@
...
 
 use std::error::Error as StdError;
 use std::fmt;
 use std::sync::Arc;
 use std::time::SystemTimeError;
 
+/// A corrupt TLS message payload that resulted in an error.
+#[non_exhaustive]
+#[derive(Debug, Clone)]
+pub enum CorruptMessageError {
```
[the remark this answers is not in the document]
Comment (complexspaces) [reply rustls/src/error.rs]: I switched this to `Invalid`, since I agree its more generic/future-proof.

File: rustls/src/error.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (djc) [reply rustls/src/error.rs]: I think that's actually fairly common, but I don't care a lot.

File: rustls/examples/internal/bogo_shim.rs [hunk 6]
```diff
@@ -9,6 +9,7 @@ use env_logger;
 use rustls;
 
 use rustls::internal::msgs::codec::Codec;
+use rustls::internal::msgs::enums::KeyUpdateRequest;
```
Comment (djc) [inline rustls/examples/internal/bogo_shim.rs:12]: This seems unrelated?

File: rustls/examples/internal/bogo_shim.rs [hunk 7]
```diff
@@ -764,7 +781,8 @@ fn exec(opts: &Options, mut sess: Connection, count: usize) {
             let mut one_byte = [0u8];
             let mut cursor = io::Cursor::new(&mut one_byte[..]);
             sess.write_tls(&mut cursor).unwrap();
-            conn.write(&one_byte).expect("IO error");
+            conn.write_all(&one_byte)
```
Comment (djc) [inline rustls/examples/internal/bogo_shim.rs:784]: Unrelated?

File: rustls/examples/internal/bogo_shim.rs [hunk 8]
```diff
@@ -610,13 +611,29 @@ fn handle_err(err: rustls::Error) -> ! {
...
+        | Error::CorruptMessage(InvalidMessage::InvalidDhParams)
+        | Error::CorruptMessage(InvalidMessage::MissingKeyExchange) => quit(":BAD_HANDSHAKE_MSG:"),
+        Error::CorruptMessage(InvalidMessage::InvalidContentType)
+        | Error::CorruptMessage(InvalidMessage::EmptyDataForDisallowedRecord)
+        | Error::CorruptMessage(InvalidMessage::UnknownProtocolVersion)
+        | Error::CorruptMessage(InvalidMessage::MessageTooLarge)
+        | Error::CorruptMessage(InvalidMessage::MissingPayload(ContentType::Unknown(42))) => {
+            quit(":GARBAGE:")
+        }
+        Error::CorruptMessage(InvalidMessage::IncorrectFrame) => quit(":GARBAGE:"),
```
Comment (djc) [inline rustls/examples/internal/bogo_shim.rs:636]: Can we do a catchall `Error::CorruptMessage(_)` here at some point so we don't have to exhaustively list all the cases manually?

File: rustls/src/client/tls12.rs [hunk 9]
```diff
@@ -405,7 +405,7 @@ impl State<ClientConnectionData> for ExpectServerKx {
             .ok_or_else(|| {
                 cx.common
                     .send_fatal_alert(AlertDescription::DecodeError);
-                Error::CorruptMessagePayload(ContentType::Handshake)
+                Error::CorruptMessage(InvalidMessage::MissingKeyExchange)
```
Comment (djc) [inline rustls/src/client/tls12.rs:408]: Similar to the earlier PRs, it would good to add some `From` impls to make this slightly less wordy. For example, if there was an `impl From<InvalidMessage> for Error` then we could probably just yield `InvalidMessage::MissingKeyExchange` here, and let the `?` operator take care of wrapping it in `Error`?

File: rustls/src/client/tls13.rs [hunk 10]
```diff
@@ -522,7 +522,7 @@ impl State<ClientConnectionData> for ExpectCertificateRequest {
             warn!("Server sent non-empty certreq context");
             cx.common
                 .send_fatal_alert(AlertDescription::DecodeError);
-            return Err(Error::CorruptMessagePayload(ContentType::Handshake));
+            return Err(Error::CorruptMessage(InvalidMessage::InvalidCertRequest));
```
Comment (djc) [inline rustls/src/client/tls13.rs:525]: Same here with the explicit wrapping.

File: rustls/src/error.rs [hunk 11]
```diff
@@ -1,11 +1,40 @@
-use crate::msgs::enums::{AlertDescription, ContentType, HandshakeType};
+use crate::msgs::enums::{AlertDescription, ContentType, HandshakeType, KeyUpdateRequest};
 use crate::rand;
 
 use std::error::Error as StdError;
 use std::fmt;
 use std::sync::Arc;
 use std::time::SystemTimeError;
 
+/// A corrupt TLS message payload that resulted in an error.
```
Comment (djc) [inline rustls/src/error.rs:9]: Style nit: probably better to move this below the `Error` definition, close to `PeerMisbehaved`.

File: rustls/src/msgs/deframer.rs [hunk 12]
```diff
@@ -70,12 +70,22 @@ impl MessageDeframer {
             let mut rd = codec::Reader::init(&self.buf[start..self.used]);
             let m = match OpaqueMessage::read(&mut rd) {
                 Ok(m) => m,
-                Err(MessageError::TooShortForHeader | MessageError::TooShortForLength) => {
-                    return Ok(None)
-                }
-                Err(_) => {
+                Err(msg_err) => {
```
Comment (djc) [inline rustls/src/msgs/deframer.rs:73]: Why change this to move the `TooShortForHeader | TooShortForLength` arms into a nested `match`? That doesn't seem like an improvement to me, since they're handled differently than all the the other cases.

Hmm, I guess it has to do with the exhaustiveness of matching here? I guess it's a little ugly but I can't quite think of a better solution.

File: rustls/src/msgs/message.rs [hunk 13]
```diff
@@ -91,23 +92,23 @@ impl OpaqueMessage {
         //  implemented per section 5.1 of RFC8446 (TLSv1.3)
         //              per section 6.2.1 of RFC5246 (TLSv1.2)
         if typ != ContentType::ApplicationData && len == 0 {
-            return Err(MessageError::IllegalLength);
+            return Err(MessageError::EmptyDataForDisallowedRecord);
```
Comment (djc) [inline rustls/src/msgs/message.rs:95]: This variant name doesn't really make sense to me, what are you trying to convey here? It seems to be much more abstract than the `EmptyApplicationData` I proposed.

File: rustls/src/tls12/mod.rs [hunk 14]
```diff
@@ -1,14 +1,14 @@
...
-use crate::kx;
 use crate::msgs::codec::{Codec, Reader};
-use crate::msgs::enums::{AlertDescription, ContentType};
+use crate::msgs::enums::AlertDescription;
 use crate::msgs::handshake::KeyExchangeAlgorithm;
 use crate::suites::{BulkAlgorithm, CipherSuiteCommon, SupportedCipherSuite};
 #[cfg(feature = "secret_extraction")]
 use crate::suites::{ConnectionTrafficSecrets, PartiallyExtractedSecrets};
 use crate::Error;
+use crate::{kx, InvalidMessage};
```
Comment (djc) [inline rustls/src/tls12/mod.rs:11]: Style nit: I would prefer to live `crate::kx` as is and import `Error` and `InvalidMessage` both directly from `crate::error`.

File: rustls/src/error.rs [hunk 15]
```diff
@@ -1,11 +1,40 @@
...
+#[non_exhaustive]
+#[derive(Debug, Clone)]
```
Comment (djc) [inline rustls/src/error.rs:11]: I think we want to derive `Copy` here, too.

File: rustls/src/msgs/deframer.rs [hunk 16]
```diff
@@ -84,8 +84,11 @@ impl MessageDeframer {
                             InvalidMessage::UnknownProtocolVersion
                         }
                     };
-                    self.desynced = true;
-                    return Err(Error::CorruptMessage(err_kind));
+
+                    return Err(self
+                        .last_error
+                        .insert(Error::CorruptMessage(err_kind))
+                        .clone());
```
Comment (djc) [inline rustls/src/msgs/deframer.rs:91]: Hmm, I would maybe go with `self.last_error = Some(Error::from(err_kind)); return Err(Error::from(err_kind));`, preferring the slight duplication over the verbosity of this construction (which is might've been nice if it didn't get spread over 4 lines).

File: rustls/src/msgs/message.rs (hunk 13 above)
Comment (complexspaces) [reply rustls/src/msgs/message.rs:95]: I'm fine working on the errors name more. I had changed it because, AFAICT, its not actually empty application data that is being checked for here. According to the RFC section (and the PR that added this, https://github.com/rustls/rustls/pull/848), application data is the only content type that's allowed to be empty, but not the others. Given that, I tried to come up with a name inverse to `EmptyApplicationData`.

File: rustls/examples/internal/bogo_shim.rs (hunk 6 above)
Comment (complexspaces) [reply rustls/examples/internal/bogo_shim.rs:12]: This is currently used in the error handling shim further down in the file:
```rust
Error::CorruptMessage(InvalidMessage::InvalidKeyUpdate(KeyUpdateRequest::Unknown(42))) => {
            quit(":BAD_HANDSHAKE_MSG:")
        }
```

Would you prefer that I removed it in favor of a less explicit `InvalidMessage::InvalidKeyUpdate(_)` pattern instead?

File: rustls/examples/internal/bogo_shim.rs (hunk 7 above)
Comment (complexspaces) [reply rustls/examples/internal/bogo_shim.rs:784]: I didn't change this manually. For some reason, running `cargo fmt` keeps forcing it onto a second line on my branch.

File: rustls/src/error.rs (hunk 15 above)
Comment (complexspaces) [reply rustls/src/error.rs:11]: Sure, I can do that if you'd like. I originally withheld from doing so since I was matching the traits implemented by `PeerIncompatible`, which only has `Clone`

File: rustls/src/msgs/deframer.rs (hunk 12 above)
Comment (complexspaces) [reply rustls/src/msgs/deframer.rs:73]: Yes, that's the reason why 👍. There's no way to tell Rust you handled some cases in a different early return branch and I figured this was more preferable to using `unreachable!()`.

File: rustls/src/msgs/deframer.rs (hunk 16 above)
Comment (complexspaces) [reply rustls/src/msgs/deframer.rs:91]: Does this feedback apply to all callsites that are currently using the `.insert()` pattern? I had originally done it with an intermediate everywhere but it got kind of cluttered.

Comment (djc) [reply rustls/src/msgs/deframer.rs:91]: Hmm, why was it cluttered? Maybe one we derive `Copy` we can away without the `.clone()` call (might need a deref operator) so things become a little cleaner?

File: rustls/src/msgs/message.rs (hunk 13 above)
Comment (djc) [reply rustls/src/msgs/message.rs:95]: Ah yes, sorry, I missed that. How about `InvalidEmptyPayload`?

File: rustls/src/msgs/deframer.rs (hunk 12 above)
Comment (djc) [reply rustls/src/msgs/deframer.rs:73]: Okay, let's leave it like this, I guess.

File: rustls/src/error.rs [hunk 17]
```diff
@@ -1,11 +1,40 @@
...
+pub enum InvalidMessage {
+    /// The peer sent us a syntactically incorrect TLS message.
```
Comment (nitnelave) [inline rustls/src/error.rs:13]: Just from a user's point of view, what would be the error message when sending a starttls message to a TLS server or vice-versa? Is there a way to make it painfully clear in the error message?

**Comment by djc** [comment]
Maybe also add a commit somewhat like this:

```diff
diff --git a/rustls/src/msgs/message.rs b/rustls/src/msgs/message.rs
index 6d1b2e60..861b47e0 100644
--- a/rustls/src/msgs/message.rs
+++ b/rustls/src/msgs/message.rs
@@ -84,26 +84,12 @@ impl OpaqueMessage {
     /// become valid if we read more data) and invalid data.
     pub fn read(r: &mut Reader) -> Result<Self, MessageError> {
         let typ = ContentType::read(r).ok_or(MessageError::TooShortForHeader)?;
-        let version = ProtocolVersion::read(r).ok_or(MessageError::TooShortForHeader)?;
-        let len = u16::read(r).ok_or(MessageError::TooShortForHeader)?;
-
-        // Reject undersize messages
-        //  implemented per section 5.1 of RFC8446 (TLSv1.3)
-        //              per section 6.2.1 of RFC5246 (TLSv1.2)
-        if typ != ContentType::ApplicationData && len == 0 {
-            return Err(MessageError::IllegalLength);
-        }
-
-        // Reject oversize messages
-        if len >= Self::MAX_PAYLOAD {
-            return Err(MessageError::IllegalLength);
-        }
-
         // Don't accept any new content-types.
         if let ContentType::Unknown(_) = typ {
             return Err(MessageError::IllegalContentType);
         }
 
+        let version = ProtocolVersion::read(r).ok_or(MessageError::TooShortForHeader)?;
         // Accept only versions 0x03XX for any XX.
         match version {
             ProtocolVersion::Unknown(ref v) if (v & 0xff00) != 0x0300 => {
@@ -112,6 +98,16 @@ impl OpaqueMessage {
             _ => {}
         };
 
+        let len = u16::read(r).ok_or(MessageError::TooShortForHeader)?;
+        // Reject undersize messages
+        //  implemented per section 5.1 of RFC8446 (TLSv1.3)
+        //              per section 6.2.1 of RFC5246 (TLSv1.2)
+        if typ != ContentType::ApplicationData && len == 0 {
+            return Err(MessageError::IllegalLength);
+        } else if len >= Self::MAX_PAYLOAD { // Reject oversize messages
+            return Err(MessageError::IllegalLength);
+        }
+
         let mut sub = r
             .sub(len as usize)
             .ok_or(MessageError::TooShortForLength)?;
```

I think that would make things a bunch clearer.

File: rustls/src/error.rs (hunk 17 above)
Comment (djc) [reply rustls/src/error.rs:13]: If you send a message like `STARTTLS` to a TLS server, rustls on the server would yield an `InvaidMessage::InvalidContentType`, because it doesn't find a known content type in the range 0x14..=0x18 in the first byte. That seems pretty reasonable?

For "vice-versa", rustls wouldn't get invoked if you send TLS messages to a peer expecting a plaintext byte stream, but TLS messages are fairly recognizable from their 5 byte header, starting with the `ContentType` byte, then a two-byte version (rustls accepts 0x03XX at the message decoding level), then a two-byte length.

Comment (djc) [reply rustls/src/error.rs:13]: I don't think we have a straightforward route to decorating these error variants with more detailed context.

**Comment by djc** [comment]
@complexspaces would you have the chance to iterate on this soon? Would you like me to take over?

**Comment by complexspaces** [comment]
Sorry for the delay, @djc. I'll push the next iteration tomorrow.

**Comment by djc** [comment]
No need to apologize, I can see how this PR might not be your top priority in life. 😉

File: rustls/src/error.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (complexspaces) [reply rustls/src/error.rs]: I didn't really feel strongly on this, so I've renamed this to `InvalidMessage` to match.

File: rustls/examples/internal/bogo_shim.rs (hunk 8 above)
Comment (complexspaces) [reply rustls/examples/internal/bogo_shim.rs]: This is now done 👍

File: rustls/src/client/tls12.rs (hunk 9 above)
Comment (complexspaces) [reply rustls/src/client/tls12.rs]: I modified all of the inline/chained occurrences of this pattern. I left the ones that had just error construction in an un-nested line alone for clarity.

File: rustls/src/error.rs (hunk 11 above)
Comment (complexspaces) [reply rustls/src/error.rs]: Good call, this has been moved down in the file.

File: rustls/src/msgs/message.rs (hunk 13 above)
Comment (complexspaces) [reply rustls/src/msgs/message.rs]: That works for me. I think its more clear when combined with the doc comments then what I had.

File: rustls/src/msgs/deframer.rs (hunk 16 above)
Comment (complexspaces) [reply rustls/src/msgs/deframer.rs]: I ended up trying out a very small helper function, `set_err`, that encloses this convert, assign, return pattern the deframer uses in multiple places. The shortness of the function allowed all of the callsites to be formatted pretty cleanly, IMO. Please take a look and see if that also works for you.

**Review by djc — APPROVED** [review]

**Comment by complexspaces** [comment]
Sorry about the import mess, @djc. I missed those when going through this branch again.

**Comment by djc** [comment]
No problem!
